desc/equilibrium/equilibrium.py

```diff
@@ -843,60 +873,128 @@ def compute(
         p = "desc.equilibrium.equilibrium.Equilibrium"
-        # If the user wants to compute x which depends on y which in turn depends on z,
-        # and they pass in y already computed in data, then we shouldn't need to compute
-        # z at all.
-        deps = list(
-            set(get_data_deps(names, obj=p, has_axis=grid.axis.size) + names)
-            - data.keys()  # subtract out y if already computed
+        deps = set(
+            get_data_deps(names, obj=p, has_axis=grid.axis.size, data=data) + names
         )
-        # TODO: replace this logic with `grid_type` from data_index
-        dep0d = [
-            dep
-            for dep in deps
-            if (data_index[p][dep]["coordinates"] == "") and (dep not in data)
-        ]
-        dep1dr = [
+
+        def need_src(name):
+            # Need to compute these on grid that is paired to the source grid, since
+            # the compute logic assume input data is evaluated on those coordinates.
+            # We exclude these from the depXdx sets below since the grids we will
+            # use to compute those dependencies are coordinate-blind.
+            # Example, "<L|r,a>" has coordinates="r", but requires computing on
+            # field line following source grid.
+            return bool(data_index[p][name]["source_grid_requirement"])
+
+        # Need to call _grow_seeds so that some other quantity like K = 2 * <L|r,a>,
+        # which does not need a source grid to evaluate, does not compute <L|r,a> on a
+        # grid that does not follow field lines.
+        need_src_deps = _grow_seeds(p, set(filter(need_src, deps)), deps)
+
+        dep0d = {
+            dep for dep in deps if is_0d_vol_grid(dep) and dep not in need_src_deps
+        }
+        # Unless user asks, don't try to recompute stuff which are only dependencies
+        # of dep0d. Example, suppose the user supplied grid is a field-line following
+        # grid, and the user would like to compute the effective ripple, which requires
+        # the scalar R0 as a dependency. The scalar R0 has the following dependencies:
+        # R0 <- A <- A(z). Each of these are computable on the quadrature grid, and
+        # since R0 is a scalar we can trivially interpolate it back to the user-supplied
+        # grid. We don't need to additionally compute A(z) and interpolate it back;
+        # it was only needed to compute R0, so we should remove it from the dep1dz list.
+        # If we don't remove it from the dep1dz list, then the code would try to create
+        # a linear grid with cross-sections at all the unique zeta values in the
+        # user-supplied grids. Typically, the user-supplied grid lacks unique_zeta_idx
+        # attribute, so this would cause an error.
+        dep0d_deps = set(
+            get_data_deps(dep0d, obj=p, has_axis=grid.axis.size, data=data)
+        )
+        # This filter is stronger than the name implies, but the false positives
```

Comment:
It may return `True` for some quantities that are not just dependencies of 0d dependencies, in other words it can generate a false positive. This is fine though since the false positive that is removed must necessarily be  computed when the actual computation happens.

For example, say we want to compute
1. A flux surface quantity `F` where `C` is a dependency of `F`.
2.  and a volume quantity `V`  where `C` is a dependency of `V`.

Now the user calls `eq.compute(names=[F, V]`, and we gather everything that needs to be compute: `[F, C, V]`. This "filter" will notice that `C` is in the dependencies for volume quantity `V` and not in `names` and then it will be removed from `dep1r`, the list of stuff that should be computed on grid's with resolution over flux surfaces. Here `C` was not `just_dep0d_dep`, but it's okay to take it out of `dep1dr` because the compute function will compute `C`  before `F` is computed anyway.